src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -0,0 +1,339 @@
+        }
+
+        private void PrepareForVerify(ReadOnlySpan<byte> content, out int alg, out byte[] toBeSigned)
+        {
+            ThrowIfUnsupportedHeaders();
+
+            ReadOnlyMemory<byte> encodedAlg = GetCoseAlgorithmFromProtectedHeaders(ProtectedHeaders);
+
+            int? nullableAlg = DecodeCoseAlgorithmHeader(encodedAlg);
+            if (nullableAlg == null)
+            {
+                throw new CryptographicException(SR.Sign1VerifyAlgHeaderWasIncorrect);
+            }
+
+            alg = nullableAlg.Value;
+            toBeSigned = CreateToBeSigned(SigStructureCoxtextSign1, _protectedHeaderAsBstr, content);
+        }
+
+        private static ReadOnlyMemory<byte> GetCoseAlgorithmFromProtectedHeaders(CoseHeaderMap protectedHeaders)
+        {
+            // https://datatracker.ietf.org/doc/html/rfc8152#section-3.1 alg:
+            // This parameter MUST be authenticated where the ability to do so exists.
+            // This authentication can be done either by placing the header in the protected header bucket or as part of the externally supplied data.
+            if (!protectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out ReadOnlyMemory<byte> encodedAlg))
+            {
+                throw new CryptographicException(SR.Sign1SignAlgIsRequired);
+            }
+
+            return encodedAlg;
+        }
+
+        // If we Validate: The caller did specify a COSE Algorithm, we will make sure it matches the specified key and hash algorithm.
+        // If we Add: The caller did not specify a COSE Algorithm, we will add the header for them, rather than throw. protectedHeaders will be altered in this case.
+        private static void ValidateOrAddAlgorithmHeader(CoseHeaderMap protectedHeaders, CoseHeaderMap unprotectedHeaders, KeyType keyType, HashAlgorithmName hashAlgorithm)
+        {
+            int algHeaderValue;
+
+            if (protectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out ReadOnlyMemory<byte> encodedAlg))
+            {
+                int? algHeaderValueNullable = DecodeCoseAlgorithmHeader(encodedAlg);
+                Debug.Assert(algHeaderValueNullable.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
+                algHeaderValue = algHeaderValueNullable.Value;
+
+                HashAlgorithmName expectedHashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(algHeaderValue, keyType);
+
+                if (expectedHashAlgorithm != hashAlgorithm)
+                {
+                    throw new CryptographicException(SR.Format(SR.Sign1SignHashAlgorithmDoesNotMatchSpecifiedAlg, hashAlgorithm.Name, algHeaderValue));
+                }
+            }
+
+            if (unprotectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out _))
+            {
+                throw new CryptographicException(SR.Sign1SignAlgIsRequired);
```

Comment:
This resource string (and identifier) says the header is required. But that's no longer true.

Reply:
This message was being shared with Verify, which still validates that the alg header is in the protected headers. 
I will add a separate string for this case, thanks.